src/libmongoc/tests/test-mongoc-loadbalanced.c

```diff
@@ -20,6 +20,127 @@
+   int topology_opening_events;
+   int topology_closed_events;
+} stats_t;
+
+static void
+server_changed (const mongoc_apm_server_changed_t *event)
+{
+   stats_t *context;
+
+   context = (stats_t *) mongoc_apm_server_changed_get_context (event);
+   context->server_changed_events++;
+}
+
+
+static void
+server_opening (const mongoc_apm_server_opening_t *event)
+{
+   stats_t *context;
+
+   context = (stats_t *) mongoc_apm_server_opening_get_context (event);
+   context->server_opening_events++;
+}
+
+
+static void
+server_closed (const mongoc_apm_server_closed_t *event)
+{
+   stats_t *context;
+
+   context = (stats_t *) mongoc_apm_server_closed_get_context (event);
+   context->server_closed_events++;
+}
+
+
+static void
+topology_changed (const mongoc_apm_topology_changed_t *event)
+{
+   stats_t *context;
+
+   context = (stats_t *) mongoc_apm_topology_changed_get_context (event);
+   context->topology_changed_events++;
+}
+
+
+static void
+topology_opening (const mongoc_apm_topology_opening_t *event)
+{
+   stats_t *context;
+
+   context = (stats_t *) mongoc_apm_topology_opening_get_context (event);
+   context->topology_opening_events++;
+}
+
+
+static void
+topology_closed (const mongoc_apm_topology_closed_t *event)
+{
+   stats_t *context;
+
+   context = (stats_t *) mongoc_apm_topology_closed_get_context (event);
+   context->topology_closed_events++;
+}
+
+static mongoc_apm_callbacks_t *
+make_callbacks (void)
+{
+   mongoc_apm_callbacks_t *cbs;
+
+   cbs = mongoc_apm_callbacks_new ();
+   mongoc_apm_set_server_changed_cb (cbs, server_changed);
+   mongoc_apm_set_server_opening_cb (cbs, server_opening);
+   mongoc_apm_set_server_closed_cb (cbs, server_closed);
+   mongoc_apm_set_topology_changed_cb (cbs, topology_changed);
+   mongoc_apm_set_topology_opening_cb (cbs, topology_opening);
+   mongoc_apm_set_topology_closed_cb (cbs, topology_closed);
+   return cbs;
+}
+
+static stats_t *
+set_client_callbacks (mongoc_client_t *client)
+{
+   mongoc_apm_callbacks_t *cbs;
+   stats_t *stats;
+
+   stats = bson_malloc0 (sizeof (stats_t));
+   cbs = make_callbacks ();
+   mongoc_client_set_apm_callbacks (client, cbs, stats);
+   mongoc_apm_callbacks_destroy (cbs);
+   return stats;
+}
+
+static stats_t *
+set_client_pool_callbacks (mongoc_client_pool_t *pool)
+{
+   mongoc_apm_callbacks_t *cbs;
+   stats_t *stats;
+
+   stats = bson_malloc0 (sizeof (stats_t));
+   cbs = make_callbacks ();
+   mongoc_client_pool_set_apm_callbacks (pool, cbs, stats);
+   mongoc_apm_callbacks_destroy (cbs);
+   return stats;
+}
+
+static void
+free_and_assert_stats (stats_t *stats)
+{
+   ASSERT_CMPINT (stats->topology_opening_events, ==, 1);
+   ASSERT_CMPINT (stats->topology_changed_events, ==, 2);
+   ASSERT_CMPINT (stats->server_opening_events, ==, 1);
+   ASSERT_CMPINT (stats->server_changed_events, ==, 1);
+   ASSERT_CMPINT (stats->server_closed_events, ==, 1);
+   ASSERT_CMPINT (stats->topology_closed_events, ==, 1);
+   bson_free (stats);
```

Comment:
Why are these hardcoded assertions? Do all the below tests have the same SDAM events?

Reply:
> Why are these hardcoded assertions?

This is to guarantee that exactly just the SDAM events for load balancer are emitted, and nothing additional. I thought this was valuable since:

1. Single-threaded connection establishment goes through SDAM machinery and could risk emitting additional ServerChanged / TopologyChanged events.
2. Pooled connections, by default, update the topology on handshake.

> Do all the below tests have the same SDAM events?

Ah! I missed this on `test_loadbalanced_server_selection_establishes_connection_single`. Added.